src/passkey_child/passkey_child_devices.c

```diff
@@ -55,32 +55,190 @@ list_devices(fido_dev_info_t *dev_list, size_t *dev_number)
+    }
+
+    ret = EOK;
+
+done:
+    if (ret != EOK) {
+        if (dev != NULL) {
+            fido_dev_close(dev);
+        }
+        fido_dev_free(&dev);
+    }
+
+    return ret;
+}
+
+errno_t
+select_from_multiple_devices(fido_dev_info_t *dev_list,
+                             size_t dev_list_len,
+                             fido_assert_t *assert,
+                             fido_dev_t **_dev)
+{
+    fido_dev_t *dev = NULL;
+    const fido_dev_info_t *di = NULL;
+    const char *path;
+    bool is_fido2;
+    errno_t ret;
+
+    DEBUG(SSSDBG_TRACE_FUNC,
+          "Working with %ld authenticator(s).\n", dev_list_len);
+
+    for (size_t i = 0; i < dev_list_len; i++) {
+        dev = fido_dev_new();
+        if (dev == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_dev_new failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        di = fido_dev_info_ptr(dev_list, i);
+        if (di == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_dev_info_ptr failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        path = fido_dev_info_path(di);
+        if (path == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_dev_info_path failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+
+        ret = fido_dev_open(dev, path);
+        if (ret != FIDO_OK) {
+            DEBUG(SSSDBG_OP_FAILURE, "fido_dev_open failed [%d]: %s.\n",
+                ret, fido_strerr(ret));
+        }
+
+        is_fido2 = fido_dev_is_fido2(dev);
+        ret = fido_dev_get_assert(dev, assert, NULL);
+        if ((is_fido2 == false && ret == FIDO_ERR_USER_PRESENCE_REQUIRED)
+            || (is_fido2 == true && ret == FIDO_OK)) {
+            *_dev = dev;
+            DEBUG(SSSDBG_FUNC_DATA, "Assertion found in passkey %ld.\n", i);
+            ret = EOK;
+            goto done;
+        }
+
+        DEBUG(SSSDBG_FUNC_DATA, "Assertion not found in passkey %ld.\n", i);
+
+        fido_dev_close(dev);
+        fido_dev_free(&dev);
+    }
+
+    ret = FIDO_ERR_NOTFOUND;
+    DEBUG(SSSDBG_OP_FAILURE, "Assertion not found.\n");
+
+done:
+    return ret;
+}
+
+errno_t
+get_device_options(fido_dev_t *dev, struct passkey_data *_data)
+{
+    bool has_pin;
+    bool has_uv;
+    errno_t ret;
+
+    has_uv = fido_dev_has_uv(dev);
+    has_pin = fido_dev_has_pin(dev);
+
+    if (_data->user_verification == FIDO_OPT_TRUE && has_pin != true
+        && has_uv != true) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Policy enabled user-verification but there isn't any "
+              "verification method set.\n");
+        ret = EINVAL;
+        goto done;
+    } else if (_data->user_verification == FIDO_OPT_TRUE && has_uv != true) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Policy enabled user-verification but the key settings disabled "
+              "it. Thus, omitting it.\n");
+        _data->user_verification = FIDO_OPT_OMIT;
+        ret = EOK;
         goto done;
     }
```

Comment:
I just had a closer look at the libfido2 behavior: You can safely drop this else-if block if you want to clean up your code slightly. Since you're guarding the built-in UV assertion request with with `fido_dev_has_uv()`, libfido2 will performs this omission for you internally for assertions with a PIN. Sorry for all the back and forth regarding this particularity.

Reply:
In that case I remove this part of the code.